daemon: Add ingress endpoint

jrajahalme opened the pull request:
Add Cilium Endpoint representing Cilium Ingress. It is defined without a veth interface so no bpf programs or maps are created for it. Ingress endpoint is needed so that the network policy is computed and configured to Envoy, so that ingress/egress network policy defined for Ingress can be enforced.

Cilium Ingress is implemented as L7 LB, which is an Envoy redirect on the egress packet path. Egress CNP policies are already enforced when defined. Prior to this commit CNPs defined for reserved:ingress identity were not computed, however, and all traffic was passed through by Cilium Ingress was allowed to egress towards the backends. When the backends receive such packets, they are identified as coming from Cilium Ingress, so any ingress policies at the backends can not discern the original source of the traffic.

This PR adds a Cilium endpoint with the reserved:ingress identity, which makes the Cilium node to compute and pass policies whose endpoint selector selects this identity (e.g., by selecting all entities) to Envoy, so that they can be enforced. Envoy listener will then enforce not just the egress policy but also the ingress policy for the original incoming source security identity.

```release-note
Network policies for reserved:ingress identity are now enforced by Cilium Ingress and Gateway API.
```

jrajahalme committed:
envoy: Update to cilium-envoy with enhanced L7 LB policy enforcement

Update cilium-envoy image to main that now has enhanced support for L7 LB
policy enforcement.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Fix support for allow-all network policies

Generate the ingress/egress network policy also when l4 filter is
nil. This enables creating allow-all rules when policy is not enforced.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

kaworu reviewed (approved):
Helm changes LGTM

jrajahalme committed:
envoy: Set enforce_policy_on_l7lb on l7 LB listener filters

Turn on ingress policy enforcement on L7 LB. With this cilium-envoy
starts dropping Ingress traffic unless Cilium Agent configures it with a
passing policy via the Ingress endpoint (with the reserved:ingress
identity).

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
k8s: Use reserved:ingress identity also for Gateway API

Tell envoy package to use the IPs allocated for reserved:ingress identity
also when a CEC/CCEC is owned by Gateway API in addition to Cilium
Ingress.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1755,6 +1755,25 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+			// allocated first. This happens earlier in the agent bootstrap.
+			if (option.Config.EnableIPv4 && len(node.GetIngressIPv4()) == 0) ||
+				(option.Config.EnableIPv6 && len(node.GetIngressIPv6()) == 0) {
+				log.Info("Ingress IPs are not available, skipping creation of the Inress Endpoint")
```
`s/Inress/Ingress`

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1755,6 +1755,25 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+			// allocated first. This happens earlier in the agent bootstrap.
+			if (option.Config.EnableIPv4 && len(node.GetIngressIPv4()) == 0) ||
+				(option.Config.EnableIPv6 && len(node.GetIngressIPv6()) == 0) {
+				log.Info("Ingress IPs are not available, skipping creation of the Inress Endpoint")
```
As a user this info message is not useful to me. Ok so "Ingress endpoint" is not created, now what? Will Cilium function properly? If no, then why don't we error / fatal instead? If Cilium will function properly then why is it important to create an Ingress Endpoint?

jrajahalme replied:
@aanm Changed this to:
```
				log.Warn("Ingress IPs are not available, skipping creation of the Ingress Endpoint: Policy enforcement on Cilium Ingress will not work as expected.")
```

Does this resolve your concern?

joestringer replied:
This doesn't resolve the concern from my perspective. Andre asked questions that need to be considered and discussed, changing a log level does not complete the discussion.

jrajahalme replied:
This discussion was completed in Slack after the fact. In summary, there is no policy breach if the Ingress Endpoint is not created, but Cilium Ingress traffic will be dropped if that is the case.

aanm reviewed (approved):
LGTM just one small comment.

Can we have cilium-cli connectivity tests that test this functionality?

bimmlerd commented on pkg/node/ingress_endpoint.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+var (
+	ingressEndpointID = templateIngressEndpointID
```
Let's avoid adding more globals here, as they make modularisation and testing hard. 

Take a look at `local_node_store.go` and it's `LocalNode` struct to see how to structure the code in a way that avoids globals; ideally the ingress endpoint functionality directly depends on and interacts with the `LocalNodeStore`, without going through some `pkg/node` function.

jrajahalme replied:
Added a commit to remove the global `node.endpointID` that I had used as the pattern for the `node.ingressEndpointID`, then changed the ingress endpoint ID use the same pattern.

jrajahalme replied:
@bimmlerd After Jussi's pushback on the approach I took for the above I noticed that the node package getter I added was not actually used at all... so I removed the setter and the global variable as well.

The already existing host endpoint ID getter is used to feed it into bpf datapath; it is currently based on a global variable in the node package.

tklauser reviewed (approved):
Looks good for @cilium/vendor

jrajahalme committed:
daemon: Add ingress endpoint

Add Cilium Endpoint representing Ingress. It is defined without a veth
interface and no bpf programs or maps are created for it. Ingress
endpoint is needed so that the network policy is computed and configured
to Envoy, so that ingress/egress network policy defined for Ingress can
be enforced.

Cilium Ingress is implemented as L7 LB, which is an Envoy redirect on the
egress packet path. Egress CNP policies are already enforced when
defined. Prior to this commit CNPs defined for reserved:ingress identity
were not computed, however, and all traffic was passed through by Cilium
Ingress was allowed to egress towards the backends. When the backends
receive such packets, they are identified as coming from Cilium Ingress,
so any ingress policies at the backends can not discern the original
source of the traffic.

This commit adds a Cilium endpoint for the reserved:ingress identity,
which makes the Cilium node compute and pass policies whose endpoint
selector selects this identity (e.g., by selecting all entities) to
Envoy, so that they can be enforced. Envoy listener will then enforce not
just the egress policy but also the ingress policy for the original
incoming source security identity.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joamaki commented on pkg/node/local_node_store.go:
```diff
@@ -18,6 +18,30 @@ type LocalNode struct {
+}
+
+// GetEndpointID returns the ID of the host endpoint for the LocalNode
+func (n *LocalNode) GetEndpointID() uint64 {
```
Since we don't rely anywhere on having to satisfy an interface with these methods, I would do the normal thing and just export `HostEndpointID` and `IngressEndpointID` and not do java-like getters.

jrajahalme replied:
Point taken, except for being moot for now (see below).

joamaki commented on pkg/node/ingress_endpoint.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+// GetIngressEndpointID returns the ID of the ingress endpoint for this node.
+func GetIngressEndpointID() uint64 {
```
Are these really needed? Can't you just take `LocalNodeStore` in where it's needed and access it through that? We are trying to get rid of these global getters and setters so it's not great that new ones are added.

jrajahalme replied:
Good question, not really! I had used the pattern of host EndpointID in the node package, but only now noticed that it is needed for feeding the host endpoint ID to the bpf datapath; there is no need for that for ingress endpoint ID.

joamaki commented on pkg/endpointmanager/manager.go:
```diff
@@ -667,6 +667,31 @@ func (mgr *endpointManager) AddEndpoint(owner regeneration.Owner, ep *endpoint.E
+		return err
+	}
+
+	node.SetIngressEndpointID(ep.GetID())
```
Please pass `LocalNodeStore` to endpoint manager as a dependency and call `Update` directly on it. Remove the global SetIngressEndpointID and GetIngressEndpointID.

jrajahalme replied:
I was able to remove IngressEndpointID from this PR, so this point is now moot.

joamaki commented on pkg/node/local_node_store_test.go:
```diff
@@ -20,6 +20,13 @@ type testInitializer struct{}
 	n.NodeIdentity = 1
+
+	// Set host endpoint ID to the default (unconfigured) value
+	n.SetEndpointID(TemplateHostEndpointID)
```
Why is this change to the test needed? See also my comment about using zero values for these. Having to set these manually is a footgun.

jrajahalme replied:
Removed, see following comment.

joamaki commented on pkg/node/local_node_store.go:
```diff
@@ -18,6 +18,30 @@ type LocalNode struct {
 	// encryption
 	OptOutNodeEncryption bool
+	// hostEndpointID is the endpoint ID allocated for the host
+	hostEndpointID uint64
```
To me this feels completely the wrong place to put these. I don't see any place that observes this information with `LocalNodeStore.Observe` so why is it here?

Wouldn't the endpoint manager be the more sensible place for this sort of information? It already has the `GetIngressEndpoint` anyway and it's the one that first knows the ID. Why not just store it there and provide a way of accessing it?

The `LocalNodeStore` was not meant for this sort of thing. It really was just meant for `types.Node` and not as store of random bits of information. Though on me for not documenting this well enough.

jrajahalme replied:
Luckily I noticed that I had added these IngressEndpoint ID setter/getter to the node package without actually needing them. I removed them from this PR and left the host endpoint stuff as it was (using a global variable).

joamaki commented on daemon/cmd/local_node_init.go:
```diff
@@ -78,6 +78,13 @@ func (ini *localNodeInitializer) initFromConfig(ctx context.Context, n *node.Loc
 	}
+
+	// Set host endpoint ID to the default (unconfigured) value
+	n.SetEndpointID(node.TemplateHostEndpointID)
```
Let's rely on the zero value instead and not do this. You could for example do:
```
type HostEndpointID struct {
  id uint64
  initialized bool
}

func (h *HostEndpointID) Set(id uint64) {
  h.initialized = true
  h.id = id
}

func (h *HostEndpointID) Get() uint64 {
  if !h.initialized {
    return TemplateHostEndpointID
  }
  return h.id
}
```

jrajahalme replied:
This was old code (for the host endpoint). Datapath code may depend on the current default value.

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -1764,6 +1764,25 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
 		}
 	}
 
+	if option.Config.EnableEnvoyConfig {
```
Please help with modularization and do not add new code to `daemon/cmd`.
`pkg/envoy` already is a module and thus any code related to it should be added there and not to `daemon/cmd`. This should be a start hook as part of `envoy.Cell`.

`registerEnvoyVersionCheck` could be a good template as it takes in few parameters and adds a start hook.
Are there any tricky ordering requirements for when the ingress endpoint needs to be added? E.g. is it clear what depends on these existing? If so can you think of how to model that dependency explicitly?

jrajahalme replied:
This option is not added in this PR. Refactoring the old code is beyond the scope of this PR.

joamaki replied:
I'm not talking about this option, I'm talking about the `AddIngressEndpoint` call below. That is not old code.

joamaki replied:
At least one tricky thing here is the ingress IP allocations that happen earlier here. I'm asking too much here as part of this PR! I think once ingress IP allocation is somewhere outside then we can model this also outside daemon/cmd.

joamaki commented on pkg/endpointmanager/cell.go:
```diff
@@ -74,12 +74,28 @@ type EndpointsLookup interface {
 	// AddEndpoint takes the prepared endpoint object and starts managing it.
 	AddEndpoint(owner regeneration.Owner, ep *endpoint.Endpoint, reason string) (err error)
 
+	AddIngressEndpoint(
```
A doc comment on what an ingress endpoint is would be useful.

jrajahalme replied:
Added:
```
	// AddIngressEndpoint creates an Endpoint representing Cilium Ingress on this node without a
	// corresponding container necessarily existing. This is needed to be able to ingest and
	// sync network policies applicable to Cilium Ingress to Envoy.
```

joamaki replied:
Awesome thanks!

bimmlerd reviewed (approved):
Thanks Jarno! :rocket:

sayboras reviewed (approved):
LGTM :heavy_check_mark:

joestringer commented on pkg/policy/l4.go:
```diff
@@ -1268,6 +1268,10 @@ func (l4 *L4Policy) AccumulateMapChanges(cs CachedSelector, adds, deletes []iden
 
 	for epPolicy := range users {
+		// Skip if endpoint has no policy maps
+		if !epPolicy.PolicyOwner.HasBPFPolicyMap() {
```
This seems like a break in the abstraction. Why does this deep policy calculation logic need to worry about BPF policymaps? Couldn't we just mock out the `epPolicy.policyMapChanges` object for the Ingress Endpoint so that the whole thing is a noop, instead of adding special cases like this?

jrajahalme replied:
This code path is about accumulating policy map changes, which are only used for syncing BPF policy maps. Maybe we could rename the new predicate to `HasPolicyMaps()` instead, or make `epPolicy.policyMapChanges` `nil` in this case and then check for that instead?

joestringer replied:
Hmm. If I follow right here, `users` is the list of subscribed endpoints that need to hear about incremental bpf policymap updates. If the new ingress endpoint doesn't care about these updates, why is it subscribed at all?

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -530,14 +530,31 @@ func (e *Endpoint) regenerateBPF(regenContext *regenerationContext) (revnum uint
+		}
+
+		return e.nextPolicyRevision, false, nil
+	}
```
This code is duplicate with the lines 584-594 now right? This code could really do with some structural love rather than adding additional conditional checks to further complicate the logic + duplicating logic.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -742,6 +759,30 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
+			datapathRegenCtxt.revertStack.Push(networkPolicyRevertFunc)
+		}
+		return false, nil
+	}
```
We should not be creating duplicate code paths for this special case. This style of extending this logic exacerbates the problem of why we can't test `pkg/endpoint`.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1697,6 +1719,25 @@ func (e *Endpoint) IsInit() bool {
+	defer cancel()
+	e.UpdateLabels(newCtx, epLabels, epLabels, true)
+	if errors.Is(newCtx.Err(), context.DeadlineExceeded) {
+		log.WithError(newCtx.Err()).Warning("Timed out while updating security identify for host endpoint")
```
This warning message was clearly not reviewed. `identify` vs. `identity` is easy to miss, but `host endpoint`? This is incorrect.

I was a little suspicious about the warning + no action as well, but it looks like `e.UpdateLabels()` has a controller to retry again in 5 minutes, so I think that that aspect is actually fine. I don't think the warning is necessary though because the error is already handled in `e.UpdateLabels()`.

joestringer commented on pkg/envoy/xds_server.go:
```diff
@@ -1574,17 +1574,22 @@ func getNetworkPolicy(ep endpoint.EndpointUpdater, vis *policy.VisibilityPolicy,
 		EndpointId:       ep.GetID(),
 		ConntrackMapName: ep.ConntrackNameLocked(),
 	}
-	// If no policy, deny all traffic. Otherwise, convert the policies for ingress and egress.
```
What's the testing coverage like for this? How do we know this change doesn't introduce a network policy enforcement issue?

jrajahalme replied:
This change allows policy to be sent to Envoy also when there is no policy (map) this was the case when policy was not enforced. Previously this did not matter because in such a case we never redirected any traffic to Envoy. Now, with Cilium Ingress (and L7 LB) in general, it is possible to redirect traffic to Envoy without policy being enforced. In this case we now need the "allow-all" policy to be sent to Envoy as otherwise it would drop all traffic now that we set the `EnforcePolicyOnL7Lb` flag on.

I did not add any testing for this yet, but had to add this change to be able to pass the current CI's Ingress tests. More specific Ingress policy test cases are planned to be added on the Cilium CLI test suite before the release.

joestringer commented on pkg/envoy/xds_server.go:
```diff
@@ -829,10 +829,14 @@ func getListenerFilter(isIngress bool, useOriginalSourceAddr bool, l7lb bool) *e
 		if ingressIPv4 != nil {
 			conf.Ipv4SourceAddress = ingressIPv4.String()
+			// Enforce ingress policy for Ingress
+			conf.EnforcePolicyOnL7Lb = true
```
Does this have any upgrade impacts?

jrajahalme replied:
This should not have upgrade impacts, as long as Envoy is updated at the same time, which still is the case also with the daemonset envoy mode.

joestringer reviewed (commented):
I have a few questions, primarily towards the end of my review below, around upgrade impacts, testing and potential impact on network policy enforcement behaviour.

Additionally I think we're well overdue for some proper attention to `pkg/endpoint` to start addressing tech debt in that area. I'm not a big fan of the way this significantly diverges the existing paths for core pkg/endpoint logic. The current abstractions for pkg/endpoint don't fit this new model particularly well, specifically because it assumes an endpoint is a lot more than what this new ingress endpoint is. I have concerns that between these differences and the general lack of testability in pkg/endpoint, we'll discover weird and wonderful bugs because of steps either not being taken when they should be in the ingress endpoint case, or steps being taken when they shouldn't be, and in the end the additional conditional checks make the logic more difficult to follow than before.
